### ISSUE TITLE
Capturing showToast events

### ISSUE THREAD

**Issue by sjurgis** [forcedotcom/LightningTestingService#39]
I have some logic that display showToast event for success && failure. I'd love to catch those events and so I can validate code works correctly. Is there a way to catch system events using testing service?

Alternatively, is there a Test.isRunningTest() method equivalent from Apex in LTS so I could create a workarounds for such cases as above?

**Comment by esalman-sfdc** [comment]
@sjurgis see if the ability to dynamically add event handlers to a component can help in this case. For example something like, 

Sample component Firing the Event:
`<aura:component>
    <aura:registerEvent name="showToast" type="force:showToast"/>
    <ui:button aura:id="b" label="press me" press="{!c.go}"></ui:button>
</aura:component>`

`({
    go: function(c, e, h){
        var toastEvent = $A.get("e.force:showToast");
            toastEvent.setParams({
                "title": "Success!",
                "message": "The record has been updated successfully."
            });
            toastEvent.fire();
    }
})`

Test validating by [dynamically adding a handler ](https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/js_cb_dynamic_handler.htm) to validate,

`  describe('c:egFireShowToast', function() {
        it("validate that this component fires showToast event", function(done) {
            $T.createComponent("c:egFireShowToast", null)
            .then(function(component) {
                var fired = false;
                component.addEventHandler("force:showToast", function(c, e){
                    fired = true;
                });
                component.find("b").getEvent("press").fire();;
                expect(fired).toBeTruthy();
                done();
            }).catch(function(e) {
                done.fail(e);
            });
        });
    });`

**Comment by sjurgis** [comment]
@esalman-sfdc Thanks! Looks like a good example on how to test the events.

However, I just realised e.force:showToast event is not available in Lightning Out so not really testable :(

Any ideas how I could detect test context or Lighting Out context?

**Comment by esalman-sfdc** [comment]
I don't completely understand your usecase/setup but in general its not really advisable to try and detect the context like that.

Is the component you are authoring is used inside a LightningOut App, and as far as this component is concerned it wants the container app to show a toast? Lightning Experience handles the event by default, but custom apps could have a handler as well and handle it whatever way they want.

**Comment by sjurgis** [comment]
As far as I understand LTS doesn't run in LEX but as a standalone app, so certain items are not available?

**Comment by esalman-sfdc** [comment]
yes LTS runs standalone, to enable users to focus on testing the component. Hence in the example above, if a component fires an event which LEX handles, test focusses on validating what the component does. Fact that LEX handles the event in a certain way is not in component's control anyway. 

If you have a specific usecase at hand which you aren't sure how to automate in isolation, feel free to link that here. 

You are correct in saying that certain functionality is specific to LEX, but based on what I know so far about your setup I am not sure how it relates to the original question.

**Comment by sjurgis** [comment]
Does `<aura:registerEvent name="showToast" type="force:showToast"/>` make it actually available outside LEX?

**Comment by esalman-sfdc** [comment]
yes, thats what my example above showed. For example if you have a custom application, it could add a handler by doing something like `<aura:handler event="force:showToast" ...` and respond to component's request to show a toast in whichever way it sees fit.

**Comment by ahayes91** [comment]
Probably a long shot, but is there any way to perform the following so that it is available immediately on load of the component?
```
                var fired = false;
                component.addEventHandler("force:showToast", function(c, e){
                    fired = true;
                });
```
I've got some error handling on load of the component that should show a toast message if a server call in the init method fails. I've mocked a response, but I suspect that the addEventHandler function it too late to catch the toast thrown by the init method. Any ideas?
